### DISCUSSION TITLE
Is it possible to have a websocket open and request HTTP aside?

### DISCUSSION THREAD

**Discussion by etienne-monier** [fastapi/fastapi#12144] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, WebSocket
from fastapi.responses import HTMLResponse
import asyncio

app = FastAPI()

@app.get("/items")
async def get_items():
    return [{"name": "screws"}]


@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        await websocket.send_text("Get server logs here.")
        await asyncio.sleep(0.2)
```

### Description

Hi,
I'm building an application which requires traditional HTTP requests.

Besides that, I'd like to use a websocket for a specific usage (this could be a webchat at bottom corner, but here, that's as simple sending server logs).

I've tested adding a websocket entrypoint besides my HTTP request, but opening websocket communication freezes the appli and HTTP requests do not succeed.

Is it possible to achieve such usage? By using nested applications?

As my usage is simple websocket with server events only, I thought to use server-sent events as proposed [here](https://amittallapragada.github.io/docker/fastapi/python/2020/12/23/server-side-events.html), but is freezed the HTTP API the same way.

How can I achieve this?

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.0.3

### Python Version

3.10

**Comment by YuriiMotov** [comment]
You should add `await`.
`await asyncio.sleep(0.2)` 

Without `await` this code sends `Get server logs here.` without delay and this cause freezing

**Reply by etienne-monier** [reply]
I've checked. I had a `await` in my code. This is missing in my minimal example. I added it.

**Reply by YuriiMotov** [reply]
Ok. What if you increase sleep time?

**Reply by etienne-monier** [reply]
This does not change anything

**Reply by YuriiMotov** [reply]
Actually, I just ran your code and it works well.

The problem is probably on your client's side.
Try using this code as a websocket client:

```
# pip install websocket-client

import websocket

def on_message(ws, message):
    print(f"Received message: {message}")

def on_error(ws, error):
    print(f"Encountered error: {error}")

def on_close(ws, close_status_code, close_msg):
    print("Connection closed")

def on_open(ws):
    print("Connection opened")


if __name__ == "__main__":
    ws = websocket.WebSocketApp("ws://127.0.0.1:8000/ws",
                                on_message=on_message,
                                on_error=on_error,
                                on_close=on_close)
    ws.on_open = on_open
    ws.run_forever()
```

**Reply by etienne-monier** [reply, marked as answer]
Ok, it figured out to be the iterator. My minimal example was not good. Here is my code:

```
@router.websocket("/log-ws")
async def websocket_endpoint_log(
    websocket: WebSocket, settings: Annotated[AppSettings, Depends(get_settings)]
):
    await websocket.accept()

    try:
        for line in tail("-f", str(settings.server.logging.log_file), _iter=True):
            await websocket.send_text(line)
            await asyncio.sleep(0.2)
    except Exception as e:
        _LOGGER.error(e)
    finally:
        await websocket.close()

```

The tail function comes from [sh](https://sh.readthedocs.io/) library. Yet, the `_iter` argument blocks if end of file is reached. `_iter_noblock` should be preffered. The following code works well.

```
@router.websocket("/log-ws")
async def websocket_endpoint_log(
    websocket: WebSocket, settings: Annotated[AppSettings, Depends(get_settings)]
):
    await websocket.accept()

    try:
        for line in tail("-f", str(settings.server.logging.log_file), _iter_noblock=True):
            if line != errno.EWOULDBLOCK:
                await websocket.send_text(line)
            await asyncio.sleep(0.2)
    except Exception as e:
        _LOGGER.error(e)
    finally:
        await websocket.close()
```

Sorry for the lost time :( I'm new to websockets...

**Reply by YuriiMotov** [reply]
Since this question is resolved, don't forget to close it

**Reply by etienne-monier** [reply]
Ouch! I was searching for it.
